tar/tar.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2017 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Copyright 2020 The FluxCD contributors. All rights reserved.
+// Adapted from: golang.org/x/build/internal/untar
+
+// Package tar provides ways to manage tarball files.
+package tar
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+)
+
+type Tar struct {
+	maxUntarSize int
+}
+
+// NewTar creates a new Tar.
+func NewTar() *Tar {
+	return &Tar{
+		maxUntarSize: 0, // no limit
+	}
+}
+
+// WithMaxUntarSize sets the limit size for archives being decompressed by Untar.
+func (u *Tar) WithMaxUntarSize(max int) *Tar {
+	u.maxUntarSize = max
+	return u
+}
+
+// Untar reads the gzip-compressed tar file from r and writes it into dir.
+func (u *Tar) Untar(r io.Reader, dir string) (summary string, err error) {
+	t0 := time.Now()
+	nFiles := 0
+	madeDir := map[string]bool{}
+	defer func() {
+		td := time.Since(t0)
+		if err == nil {
+			summary = fmt.Sprintf("Extracted tarball into %s: %d files, %d dirs (%v)", dir, nFiles, len(madeDir), td)
+		} else {
+			summary = fmt.Sprintf("Error extracting tarball into %s after %d files, %d dirs, %v: %v", dir, nFiles, len(madeDir), td, err)
+		}
+	}()
+	zr, err := gzip.NewReader(r)
```

Comment:
Can we have a flag to support non-compressed tarball?

Reply:
This package is mostly aimed at the Flux project needs. Is there any specific use case you have in mind in which we would need non-compressed tarballs?